src/datatypes/interval.coffee

```diff
@@ -13,6 +13,16 @@ module.exports.Interval = class Interval
     isInterval:
       get: -> true
 
+  copy: ->
+    newLow = @low
+    newHigh = @high
+    if @low.copy?()
```

Comment:
In this `if` clause (and the one below), your if condition actually _invokes_ the `copy` function -- which means that you're creating a copy (and temporarily allocating memory) just to test if the copy function exists (and if it does, then you _invoke_ copy again when you assign it).

I'd recommend you just test to see if a copy function exists like this:

```coffee
if typeof @low.copy == 'function'
```

Reply:
Same comment applies in several places throughout this PR.